prow/e2e-suite.sh

```diff
@@ -76,6 +77,9 @@ for ((i=1; i<=$#; i++)); do
     E2E_ARGS+=( "${!i}" )
 done
 
+# upload images
+time ISTIO_DOCKER_HUB="${HUB}" make push HUB="${HUB}" TAG="${TAG}"
```

Comment:
will there be problems when we try to push the same images from 20 different tests? the tag will be the same but will the images be identical?

Reply:
the image should be identical because each test will checkout the same repo SHA.
I have updated the $TAG so that it will have different name "SHA+TestName".

Reply:
Prow rebase the code from branch, and the GIT_SHA is coming directly from the PR. So it is possible that we push different code to the same image. Note that this bug was always present but since we had only on job pushing, and the other after success job were using the same refs, we were OK. (There is also the issue where a docker image depends from a moving tag like latest, in which case we cannot guarantee anything)

I think podutils will recreate SHA in deterministic way when rebasing (need to verify if the fork has it), meaning that rebase from 2 different jobs will recreated the same SHA if they rebase from the same refs. One option would be to use the Head of the rebased branch for testing. since this image will only be used in that specific job, we should be OK.

Reply:
All the jobs will be starting at the same, so even if we use the PR Head SHA, we should still be OK as those images should be the same. We need to consider the case where we retest one of the jobs in which case it will create a new image that will be used in the test. That should be OK. If that's an issue we can just use a unique identifier per test run as the docker tag.

Reply:
Prow should clone the git repo + any changes in a reproducible way now, so yes, the git sha should be consistent.

I'm not sure how your build is implemented, but if it's not reproducible (e.g. if it includes timestamps, build paths, and other things), then the image may not be identical.

Reply:
sorry, i am getting confused with this..
This is my understanding:
1. All tests will be kicked off at the same time.
2. Each test will create and use their own image (tag= sha+testname e.g: a3fd9745308c18902f78ce9ae3487296fe83edfde2e_simple)

Problem:
There is no guarantee for each test invocation's image to be identical in presubmit testing because the code might change although SHA is the same. is this right? Don't prow kicked off new tests whenever a new commit is pushed?

Above problem will only affect presubmit. There shouldn't be any problem with postsubmit testing since the code under a SHA is freezed. is this true?

Reply:
if the code changes, then the SHA will also change. all prow does is ensure that all jobs using the same code will result in the same git SHAs - e.g. if I `/retest` and nothing has merged or otherwise changed, the git SHAs should be identical. but if I push a new commit, or something is merged into master, the SHA will change.

Reply:
the images may not be identical even if the SHAs are identical, however, unless you've ensured that the builds are reproducible, which doesn't usually come automatically.

Reply:
ahh okay since we tag the images with git SHA+testName. They are all should be identical right?

Reply:
what can make images not identical when the SHAs are identical?

Reply:
A common case is when you embed timestamps into binary artifacts. Also, sometimes paths can sneak into the binaries, too - see e.g. https://blog.filippo.io/reproducing-go-binaries-byte-by-byte/.

Some of this may have improved in recent Go versions, but I know it took kubernetes several iterations to get reproducible builds (and I'm not sure it's fully there yet).

Reply:
@ixdy thank you for the info! :)

@howardjohn, is it a big problem with using non-identical images(with identical SHAs) for presubmit or postsubmit testing and is yes, why is that? We have long running integration test clusters for releases(these tests will be using daily created images)

Reply:
I think it's fine especially if we the different image names for each. I was more worried about pushing the esme tag 20 times not that the images are identical